Tests/EndToEnd.tests.ts

```diff
@@ -968,4 +976,94 @@ describe("EndToEnd", () => {
+                "vmId": "1",
+                "subscriptionId": "2",
+                "osType": "Windows_NT"
+            }`;
+            var stub: sinon.SinonStub = sandbox.stub(http, "request", (options: any, callback: any) => {
+                var req = new fakeRequest(false, "http://169.254.169.254", vmDataJSON);
+                req.on("end", callback);
+                return req;
+            });
+
+            // set up sdk
+            const client = new TelemetryClient("key");
+            const heartbeat: HeartBeat = new HeartBeat(client);
+            heartbeat.enable(true, client.config);
+            HeartBeat.INSTANCE.enable(true, client.config);
+            const trackMetricStub = sinon.stub(heartbeat["_client"], "trackMetric");
+            
+            heartbeat["trackHeartBeat"](client.config, () => {
+                assert.equal(trackMetricStub.callCount, 1, "should call trackMetric for the VM heartbeat metric");	
+                assert.equal(trackMetricStub.args[0][0].name, "HeartBeat", "should use correct name for heartbeat metric");	
+                assert.equal(trackMetricStub.args[0][0].value, 0, "value should be 0");	
+                const keys = Object.keys(trackMetricStub.args[0][0].properties);	
+                assert.equal(keys.length, 5, "should have 4 kv pairs added when resource type is VM");	
+                assert.equal(keys[0], "sdk", "sdk should be added as a key");	
+                assert.equal(keys[1], "osType", "osType should be added as a key");
+                assert.equal(keys[2], "azInst_vmId",  "azInst_vmId should be added as a key");	
+                assert.equal(keys[3], "azInst_subscriptionId", "azInst_subscriptionId should be added as a key");	
+                assert.equal(keys[4], "azInst_osType", "azInst_osType should be added as a key");	
+
+                const properties = trackMetricStub.args[0][0].properties;	
+                assert.equal(properties["sdk"], Context.sdkVersion, "sdk version should be read from Context");	
+                assert.equal(properties["osType"], os.type(), "osType should be read from os library");
+                assert.equal(properties["azInst_vmId"], "1", "azInst_vmId should be read from response");	
+                assert.equal(properties["azInst_subscriptionId"], "2", "azInst_subscriptionId should be read from response");	
+                assert.equal(properties["azInst_osType"], "Windows_NT", "azInst_osType should be read from response");
+                trackMetricStub.restore();
+                heartbeat.dispose();
+                stub.restore();
+                done();
+            });
+        });
+
+        it("should only send name and value properties for heartbeat metric when get VM request fails", (done) => {
+            // set up stub
+            var stub: sinon.SinonStub = sandbox.stub(http, "request", (options: any, callback: any) => {
+                var req = new fakeRequest(true, "http://169.254.169.254");
```

Comment:
Note the different of this req with line 1000 is req.on('end') not exist - so the response is null. and the fakeRequest fails immediately - so the get vm API throws error.